### PR TITLE
Add Functionality to Retrieve Ancestor App IDs when a Shared App is Given

### PR DESCRIPTION
**Pull request by dhaura**
### Purpose

- $Subject

### Relatad PRs
- https://github.com/wso2-extensions/identity-organization-management/pull/383

### Related Issues
- https://github.com/wso2/product-is/issues/21214

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 1]
```diff
@@ -6110,6 +6117,67 @@ public String getMainAppId(String sharedAppId) throws IdentityApplicationManagem
...
+        Map<String, String> sharedAppIds = new HashMap<>();
+
+        String placeholders = IntStream.range(0, sharedOrgIds.size())
+                .mapToObj(i -> ":" + SHARED_ORG_ID_PLACEHOLDER_PREFIX + i + ";")
+                .collect(Collectors.joining(", "));
+        String sqlStmt = GET_FILTERED_SHARED_APPLICATIONS.replace(SHARED_ORG_ID_LIST_PLACEHOLDER, placeholders);
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+
+            try (NamedPreparedStatement namedPreparedStatement =
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:6158]: add both resources into the same try block

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java [hunk 2]
```diff
@@ -511,6 +511,16 @@ public class ApplicationMgtDBQueries {
 
     public static final String GET_MAIN_APP_ID = "SELECT MAIN_APP_ID FROM SP_SHARED_APP WHERE SHARED_APP_ID = ?";
 
+    public static final String GET_OWNER_ORG_ID_BY_SHARED_APP_ID =
+            "SELECT OWNER_ORG_ID FROM SP_SHARED_APP WHERE SHARED_APP_ID = :" +
+                    SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_SHARED_APP_ID + ";";
+
+    public static final String GET_FILTERED_SHARED_APPLICATIONS =
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationMgtDBQueries.java:518]: make sure these queries work for all DB types

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 3]
```diff
@@ -6110,6 +6117,67 @@ public String getMainAppId(String sharedAppId) throws IdentityApplicationManagem
         }
     }
 
+    @Override
+    public String getOwnerOrgId(String sharedAppId) throws IdentityApplicationManagementServerException {
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (NamedPreparedStatement namedPreparedStatement = new NamedPreparedStatement(
```
[the remark this answers is not in the document]
Comment (dhaura) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/6129/commits/1c6e52945f25edf3ca8faf6da8d114cf1439aed3

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java [hunk 4]
```diff
@@ -2908,6 +2910,45 @@ public String getMainAppId(String sharedAppId) throws IdentityApplicationManagem
...
+                throw buildServerException("Error while resolving the tenant domain for the organization id: " + orgId);
+            }
+            // Check if the child app is a main application.
+            if (isMainApp(sharedAppId, tenantDomain)) {
+                return Collections.singletonMap(orgId, sharedAppId);
+            }
+            return Collections.emptyMap();
+        }
+
+        String ownerOrgId = ApplicationMgtSystemConfig.getInstance().getApplicationDAO().getOwnerOrgId(sharedAppId);
```
[the remark this answers is not in the document]
Comment (dhaura) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java:2932]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/6129/commits/1c6e52945f25edf3ca8faf6da8d114cf1439aed3

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 1 above)
Comment (dhaura) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/6129/commits/1c6e52945f25edf3ca8faf6da8d114cf1439aed3

**Review by AnuradhaSK — DISMISSED** [review]

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**